### ISSUE TITLE
Set attribute to null instead of empty list

### ISSUE THREAD

**Issue by bimworld** [IfcOpenShell/IfcOpenShell#1966]
Is it possible to write a $ (null) as attribute instead of (), if the argument to be set is an empty list? This makes a difference if cardinality rules apply.

Example:
`void setXYZ(IfcTemplatedEntityList< ::Ifc4::xyz >::ptr v);`
only allows a list as argument, so it will output () for an empty list. I then get in trouble with rule 
> OPTIONAL LIST [1:?]

**Comment by aothms** [comment]
`IfcTemplatedEntityList< ::Ifc4::xyz >::ptr` is a shared ptr, so you can probably pass `nullptr` or an uninitialized shared pointer e.g `IfcTemplatedEntityList< ::Ifc4::xyz >::ptr{}`

In 0.7 we changed this a bit. Not entirely sure in all cases, but it may be that we wrapped it in a boost::optional there. We did that at least for the simple scalar attributes.

**Comment by bimworld** [comment]
Unfortunately, both and others variations of those will segfault.

`IfcSpaceType->setRepresentationMaps(nullptr);`
`IfcSpaceType->setRepresentationMaps(typename IfcTemplatedEntityList<typename Schema::IfcRepresentationMap>::ptr{});`

This works but does ()
``` 
typename IfcTemplatedEntityList<typename Schema::IfcRepresentationMap>::ptr L(new IfcTemplatedentityList<typename Schema::IfcRepresentationMap>);  
IfcSpaceType->setRepresentationMaps(L);
```
Maybe I have to wait for 0.7 release.

**Comment by aothms** [comment]
> Maybe I have to wait for 0.7 release.

Well, you don't have to wait, you can switch to it right away, it's currently the default branch of the repository.

The type there is indeed `boost::optional< aggregate_of< ::Ifc4::IfcRepresentationMap >::ptr >`, where aggregate_of is simply a rename of IfcTemplatedEntityList

> Unfortunately, both and others variations of those will segfault.

Hm indeed, well 0.6 is more or less closed now, and it won't be a full solution, since it only concerns the aggregate arguments, but if you're interested in addressing this, you'd have to go to the code generator https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcopenshell-python/ifcopenshell/express/templates.py#L262 and change `v->generalize()`  to something like `v ? v->generalize() : (IfcEntltyList::ptr) nullptr`. I'd happily accept a PR but I'd really recommend moving to 0.7.

**Comment by Moult** [comment]
Closing as fixed in v0.7.0
